### ISSUE TITLE
No content output to blob

### ISSUE THREAD

**Issue by vplauzon** [chriswill/serilog-sinks-azureblobstorage#77]
Hi,

I'm new to Serilog & to this extension, so this is more likely a beginner's mistake than an issue with the software.

Basically I create a logger:

            var logger = new LoggerConfiguration()
                .WriteTo
                .AzureBlobStorage(
                    connectionString,
                    LogEventLevel.Verbose,
                    container,
                    "{yyyy}-{MM}-{dd}-log.txt",
                    blobSizeLimitBytes:200000000)
                .CreateLogger();

(with a valid connectionString and container)

I DO NOT set Log.Logger (I've tried with same result).

I then pass the logger instance to a service (as a singleton) and use it:

logger.Information("something");

What happens:

- No exceptions are thrown
- A blob is created in the right container 2021-03-26-log.txt at the moment the logger.Information line is executed
- (So far so good)
- **The blob is of length zero and therefore empty and remains empty**

I've tried multiple "slap the ATM" tactics such as running multiple logger.Information and never got anything in a blob.

Any tips to resolve this?

Cheers,

VP

**Comment by chriswill** [comment]
Which environment are you using for logging?  I.e. Azure App Service, Function app, or are you building some other type of app?  I don't know why this would work, but I recall one user of a Function app (I believe) had a similar issue and mentioned that they used [Serilog Async](https://github.com/serilog/serilog-sinks-async), which solved their problem.

**Comment by vplauzon** [comment]
Hi Chris,

Thanks for prompt return.

So far it is on...  my laptop ;)  But target is Azure App Service.

It is a ASP.NET Fx 5.0 project.

It is odd since it does create the blob at the right time, so obviously the pipeline doesn't hit any speed bump.

I'll give Serilog Async a shot and let you know!

Thanks.

VP

**Comment by chriswill** [comment]
Are you using the storage emulator on your laptop or an actual Azure storage account?  

The emulator does not support append blobs.

**Comment by vplauzon** [comment]
No, I am pointing directly to storage on an ADLS gen2 (i.e. hierarchical storage).

I've tried async too.  It doesn't work either...

Before throwing the towel, I've tried on a storage account without hierarchical namespace activated (i.e. not ADLS gen2) and...  it works!

...  but I need ADLS gen2...

Is that a known limitation?  I couldn't find documentation about append blob not working on ADLS gen 2 (cf https://docs.microsoft.com/en-us/azure/storage/blobs/data-lake-storage-known-issues).

I notice serilog-sinks-azureblobstorage is using Microsoft.Azure.Storage.Blob which is deprecated.  The last version dates from August 2020 and is the one used in the NuGet package...  That isn't ancient and should be compatible with ADLS gen2.

Any ideas?

**Comment by vplauzon** [comment]
Hey Chris, would you be open to a PR upgrading the Blob SDK?

I forked the repo and looked at it.  It isn't a trivial exercise as the SDK has changed quite a few esthetic things, but it wouldn't require me to understand Serilogs inside-out either.

So let me know as it would take me a few hours of work...

**Comment by chriswill** [comment]
Hi Vincent-Philippe, I'd be very open to that.  As you mentioned, the update isn't a trivial namespace update, and I just haven't had the time to start it.

Another user forked it and performed some type of upgrade as [described here](https://github.com/chriswill/serilog-sinks-azureblobstorage/issues/58#issuecomment-716152668), but they didn't submit a PR.

**Comment by vplauzon** [comment]
Ok I hit a partial wall where I would need to refactor the code substantially to resolve.  So I won't finish the package upgrade unfortunately.  Let me document to speed up the next person attempting.

The issue I found is the handling of Append blobs in the new SDK.  We need a AppendBlobClient object (e.g. https://docs.microsoft.com/en-us/dotnet/api/azure.storage.blobs.specialized.appendblobclient?view=azure-dotnet).  There is no way to create an AppendBlobClient from a BlobContainerClient (it's ironic since there is a protected method GetAppendBlobClientCore).  So the only way to create an AppendBlobClient is to construct it directly.

If we look at the constructor of AppendBlobClient, in order for it to be functional, we need to pass authentication info.  The closest matching the philosophy of the package takes a shared key credentials (cf https://docs.microsoft.com/en-us/dotnet/api/azure.storage.blobs.specialized.appendblobclient.-ctor?view=azure-dotnet#Azure_Storage_Blobs_Specialized_AppendBlobClient__ctor_System_Uri_Azure_Storage_StorageSharedKeyCredential_Azure_Storage_Blobs_BlobClientOptions_).

Technically, that is available but:

1. It isn't available at the point where we need to create the AppendBlobClient (in DefaultCloudBlobProvider).  So we would need to propagate it.
2. We receive a "connection string" from the client, so we would need to manipulate it to find the shared key.

So nothing impossible, but it does require a fair bit of work.

**Comment by vplauzon** [comment]
Strike that, there are extension methods in Azure.Storage.Blobs.Specialized.SpecializedBlobExtensions I didn't see.

**Comment by vplauzon** [comment]
Ok, I've completed the upgrade and submitted the PR at https://github.com/chriswill/serilog-sinks-azureblobstorage/pull/78.

As mentioned in the PR comment, you would likely have some work to do before committing it.

I've tested it with a console app calling a logger.information and it worked.  So pretty minimal tests!

When append-blob become GA on ADLS, this should allow us to use Serilog on ADLS gen2 :) (maybe we'll need to upgrade the SDK but by then it's going to be just a minor upgrade)
